pkg/usagetracker/tracker_test.go

```diff
@@ -3,15 +3,355 @@
+			"b1": newTestUsageTracker(t, 1, "zone-b", ikv, pkv, cluster),
+		}
+		waitUntilAllTrackersSeeAllInstancesInTheirZones(t, trackers)
+
+		// Usage-trackers have started now, we run reconciliations but nothing should happen yet
+		// because previous instances (-0) are not running yet.
+		reconcileAllTrackersPartitionCountTimes(t, trackers)
+
+		// They should not be ready yet.
+		requireAllTrackersNotReady(t, trackers)
+
+		// start zone-a-0.
+		trackers["a0"] = newTestUsageTracker(t, 0, "zone-a", ikv, pkv, cluster)
+
+		// Reconcile them again.
+		for partitionID := int32(0); partitionID < testPartitionsCount; partitionID++ {
+			for key, ut := range trackers {
+				require.NoError(t, ut.reconcilePartitions(context.Background()), "Tracker %q could not reconcile", key)
+			}
+		}
+
+		// zone-a-0 and zone-a-1 should be ready now, but zone-b-1 is not ready yet because zone-b-0 didn't join the instance ring yet.
+		require.NoError(t, trackers["a0"].CheckReady(context.Background()), "Tracker a0 should be ready now")
+		require.NoError(t, trackers["a1"].CheckReady(context.Background()), "Tracker a1 should be ready now")
+		require.Error(t, trackers["b1"].CheckReady(context.Background()), "Tracker b1 should not be ready yet")
+	})
+
+	t.Run("new instance takes partitions, old shut downs them", func(t *testing.T) {
+		t.Parallel()
+
+		// lostPartitionsShutdownGracePeriod should be long enough for this test not to be flaky,
+		// but also short enough to keep the test fast
+		const lostPartitionsShutdownGracePeriod = time.Second
+		configure := func(cfg *Config) {
+			cfg.LostPartitionsShutdownGracePeriod = lostPartitionsShutdownGracePeriod
+			cfg.MaxPartitionsToCreatePerReconcile = testPartitionsCount
+		}
+
+		ikv, pkv, cluster := prepareKVStoreAndKafkaMocks(t)
+		a0 := newTestUsageTracker(t, 0, "zone-a", ikv, pkv, cluster, configure)
+		b0 := newTestUsageTracker(t, 0, "zone-b", ikv, pkv, cluster, configure)
+		trackers := map[string]*UsageTracker{
+			"a0": a0,
+			"b0": b0,
+		}
+		waitUntilAllTrackersSeeAllInstancesInTheirZones(t, trackers)
+
+		// Create trackers and check that they're ready.
+		require.NoError(t, a0.reconcilePartitions(context.Background()))
+		require.NoError(t, b0.reconcilePartitions(context.Background()))
+		requireAllTrackersReady(t, trackers)
+
+		// Start zone-a-1.
+		a1 := newTestUsageTracker(t, 1, "zone-a", ikv, pkv, cluster, configure)
+		trackers["a1"] = a1
+
+		waitUntilAllTrackersSeeAllInstancesInTheirZones(t, trackers)
+
+		// Reconcile both, they will create all the needed partitions.
+		require.NoError(t, a0.reconcilePartitions(context.Background()))
+		require.NoError(t, a1.reconcilePartitions(context.Background()))
+
+		// All trackers should be ready.
+		requireAllTrackersReady(t, trackers)
+
+		// zone-a-0 should have 16 partitions, zone-a-1 should have 8, this is deterministic at this point.
+		a0.partitionsMtx.RLock()
+		require.Len(t, a0.partitions, 16)
```

Comment:
If this fails, the lock is not released in the next line. I quickly checked the cleanup functions and I don't think they will deadlock, but please double check it too.